tests: kernel: timer_api fails with CONFIG_STM32_LPTIM_TIMER

FRASTM opened the pull request:
Testing kernel/timer/timer_API with LPTIM on stm32L4R5
the test_timer_duration_period fails w"
test_timer_duration_period: tdata.expire_cnt == 4 is false"

- changed z_timer_cycle_get_32() function to have better value returned in nb of HW cycles
- changed the calculation for next timeout in case of short delay
- source LPTIM with LSE or LSI clock for better accuracy
- reduce tick freq per sec and adjust for a better divided value of
 LPTIM_CLOCK % CONFIG_SYS_CLOCK_TICKS_PER_SEC

Activation of the LPTIM remains an experimental feature to generate kernel timings.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/20991

FRASTM commented:
If the irq has not yet been handled but the counter just resets after the roll-over we must add thi roll-ove value in the actual counter position

FRASTM commented:
Changed the calculation of the nb of hw cycles which is returned by the **z_timer_cycle_get_32()** function in case of LSE clock.
In case of LSI clock the both methods works.

Remaining work:  set a timeout close to the current counter.

FRASTM commented:
This version passed kernel test on nulceo_l4r5zi board with LPTIM timer running on LSI clock
Device testing on:
nucleo_l4r5zi (None) on /dev/ttyACM0
Adding tasks to the queue...
total complete:   82/  82  100%  skipped:    8, failed:    0
74 of 82 tests passed (100.00%), 0 failed, 8 skipped with 0 warnings in 2287.16 seconds
In total 93 test cases were executed on 1 out of total 213 platforms (0.47%)

FRASTM commented:
as nucleo_wb55rg board has CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=32000000
a LPTIM clock source of LSI=32000 gives better accuracy in lptimer calculations

FRASTM commented:
on nucleo_wb55rg and nucleo_l4r5zi boards:
test passed with  _west build -b nucleo_wb55rg tests/kernel/fatal_
test passed with  _west build -b nucleo_wb55rg tests/kernel/sched/schedule_api_
test passed with  _west build -b nucleo_wb55rg tests/kernel/timer/_

FRASTM commented:
add change for nucleo_wb55rg board

FRASTM commented:
in the `tests/kernel/critical`, the test case regression_thread  is calling `k_sched_time_slice_set()`
	TC_PRINT("Enable timeslicing at %u\n", k_uptime_get_32());
	k_sched_time_slice_set(10, 10);
However in case of **CONFIG_TIMESLICING=n** the `k_sched_time_slice_set`  is not defined.
by the `kernel/sched.c`
Is it correct ?

FRASTM commented:
test with dummy `k_sched_time_slice_set()` function

erwango commented on boards/arm/nucleo_l4r5zi/nucleo_l4r5zi_defconfig:
```diff
@@ -12,6 +13,14 @@ CONFIG_SOC_STM32L4R5XX=y
+# CONFIG_DEBUG=y
+CONFIG_TICKLESS_KERNEL=y
+CONFIG_TIMESLICING=n
+CONFIG_SYS_CLOCK_TICKS_PER_SEC=1000
```
Do not impact default board configuration.
LPTIM conditional configuration could be done in board Kconfig.defconfig, but not here

FRASTM replied:
Enabling the LPTIMer is done in the board with 
CONFIG_STM32_LPTIM_TIMER=y

erwango commented on boards/arm/nucleo_wb55rg/nucleo_wb55rg_defconfig:
```diff
@@ -3,6 +3,14 @@ CONFIG_SOC_STM32WB55XX=y
+CONFIG_TICKLESS_KERNEL=y
+CONFIG_TIMESLICING=n
+CONFIG_SYS_CLOCK_TICKS_PER_SEC=1000
+
```
Do not impact default board configuration.
LPTIM conditional configuration could be done in board Kconfig.defconfig, but not here

FRASTM replied:
The config for STM32_LPTIM_CLOCK  is now fixed to LSE, 
the config for SYS_CLOCK_TICKS_PER_SEC is fixed 
for all socs 
in the soc/arm/st_stm32/common/Kconfig.defconfig.series

erwango commented on boards/arm/nucleo_wb55rg/nucleo_wb55rg_defconfig:
```diff
@@ -3,6 +3,14 @@ CONFIG_SOC_STM32WB55XX=y
+CONFIG_STM32_LPTIM_CLOCK_LSI=n
+# CONFIG_DEBUG=y
+CONFIG_TICKLESS_KERNEL=y
+CONFIG_TIMESLICING=n
```
Can you elaborate why you're changing these 2 settings (as I guess this is not anecdotal)

erwango commented on soc/arm/st_stm32/stm32l4/dts_fixup.h:
```diff
@@ -74,4 +74,11 @@
 #define DT_LPTIM_1_CLOCK_BITS	DT_ST_STM32_TIMERS_40007C00_CLOCK_BITS_0
 #define DT_LPTIM_1_CLOCK_BUS	DT_ST_STM32_TIMERS_40007C00_CLOCK_BUS_0
 
+#define DT_LPTIM_1_BASE_ADDRESS	DT_ST_STM32_TIMERS_40007C00_BASE_ADDRESS
```
This commit is not needed (defines are already available)

erwango commented on boards/arm/nucleo_l4r5zi/nucleo_l4r5zi_defconfig:
```diff
@@ -55,3 +55,6 @@ CONFIG_UART_CONSOLE=y
 CONFIG_ARM_MPU=y
+
+# Enable LPTimer
+CONFIG_STM32_LPTIM_TIMER=y
```
LPTIM should not be enabled by default.
Rather, it should be enabled as new configuration in timer test suites

erwango commented on soc/arm/st_stm32/stm32wb/dts_fixup.h:
```diff
@@ -43,4 +43,11 @@
 #define DT_LPTIM_1_CLOCK_BITS	DT_ST_STM32_TIMERS_40007C00_CLOCK_BITS_0
 #define DT_LPTIM_1_CLOCK_BUS	DT_ST_STM32_TIMERS_40007C00_CLOCK_BUS_0
 
+#define DT_LPTIM_1_BASE_ADDRESS	DT_ST_STM32_TIMERS_40007C00_BASE_ADDRESS
```
This commit is not needed (defines are already available)

erwango commented on boards/arm/nucleo_wb55rg/nucleo_wb55rg_defconfig:
```diff
@@ -31,3 +31,6 @@ CONFIG_UART_CONSOLE=y
 CONFIG_ARM_MPU=y
+
+# Enable LPTimer
+CONFIG_STM32_LPTIM_TIMER=y
```
LPTIM should not be enabled by default.

FRASTM commented:
removed useless constant definition in dts_fixup.h

erwango commented on soc/arm/st_stm32/common/Kconfig.defconfig.series:
```diff
@@ -12,6 +12,9 @@ config CORTEX_M_SYSTICK
 	default y if !STM32_LPTIM_TIMER
 
+config SYS_CLOCK_TICKS_PER_SEC
+	default 1024 if STM32_LPTIM_TIMER
```
Does it depends also on LPTIM clock source  (LSI or LSE)?
In that case, following would be better:
```
	default 1024 if STM32_LPTIM_TIMER AND STM32_LPTIM_CLOCK_LSE

FRASTM replied:
or even better:
```
config SYS_CLOCK_TICKS_PER_SEC
	default 1000 if STM32_LPTIM_CLOCK_LSI
	default 1024 if STM32_LPTIM_CLOCK_LSE
	depends on STM32_LPTIM_TIMER 
```

erwango replied:
Indeed!

erwango reviewed (changes requested):
New comment.
Also please review remaining comments

FRASTM commented:
fix checkpatch issues

FRASTM commented:
changing the testcases tests/kernel/timer for , nucleo_l4r5zi and nucleo_wb55rg boards
with a special config taht enables the LPTIM
TICKS_PER_SEC is set to 1000 and LSE is selected as LPTIM clock source for this tests

erwango commented on tests/kernel/timer/timer_api/testcase.yaml:
```diff
@@ -7,3 +7,10 @@ tests:
     tags: kernel userspace
+  kernel.timer.tickless.stm32:
+    filter: dt_compat_enabled("stm32,lptim")
+    platform_whitelist: nucleo_wb55rg, nucleo_l4r5zi
```
platform white listing is not needed following check on compatible

FRASTM replied:
removing the    ` platform_whitelist: nucleo_wb55rg, nucleo_l4r5zi`

erwango commented on tests/kernel/timer/timer_monotonic/testcase.yaml:
```diff
@@ -1,3 +1,9 @@
     tags: timer
+  kernel.timer.stm32:
+    filter: dt_compat_enabled("stm32,lptim")
+    platform_whitelist: nucleo_wb55rg, nucleo_l4r5zi
```
platform white listing is not needed following check on compatible

FRASTM replied:
removing the      `platform_whitelist: nucleo_wb55rg, nucleo_l4r5zi`

erwango commented on tests/kernel/timer/timer_monotonic/testcase.yaml:
```diff
@@ -1,3 +1,9 @@
 tests:
   kernel.timer.monotonic:
     tags: timer
+  kernel.timer.stm32:
```
kernel.timer.monotonic.stm32

FRASTM replied:
changing this test to `kernel.timer.monotonic.stm32`

erwango reviewed (changes requested):
Few additional comments

Also please review commit message for "soc: arm: stm32 common lptim freq and SYS_CLOCK accuracy"

FRASTM commented:
update after erwango's review

erwango commented on tests/kernel/timer/timer_api/testcase.yaml:
```diff
@@ -7,3 +7,15 @@ tests:
     platform_exclude: qemu_x86_coverage qemu_cortex_m0
     tags: kernel userspace
+  kernel.timer.tickless.stm32:
+    filter: dt_compat_enabled("stm32,lptim")
```
This compatible does not actually exists (LPTIM1 compat is "st,stm32-timers").
But indeed a dedicated lptim compatible is required so we don't mix with other general purpose timers

FRASTM replied:
adding a new compatible  = "st,stm32-lptim  especially for lptim, which is itself 'timers" compatible (9416b3d7d1ef91a2c0a3fc826e3b40ea6404871f)

erwango reviewed (changes requested):
We now need a dedicated st,stm32-lptim compatible for LPTIM1

jdascenzio committed:
driver: timer: stm32_lptim: fix autoreload value

Autoreload value must be decrement by one

Signed-off-by: Julien D'Ascenzio <julien.dascenzio@paratronic.fr>

jdascenzio committed:
driver: timer: stm32_lptim: fix hang when no tickless

When the tickless kernel isn't used, we don't want to wait for ARROK.
This wait can be endless.

Signed-off-by: Julien D'Ascenzio <julien.dascenzio@paratronic.fr>

jdascenzio committed:
driver: timer: fix accumulated counter increment

The current value of the counter must not be added to the accumulator.
It will be added when calling z_timer_cycle_get_32.

Signed-off-by: Julien D'Ascenzio <julien.dascenzio@paratronic.fr>

FRASTM commented:
Add the PR  https://github.com/zephyrproject-rtos/zephyr/pull/25126 to help fixing the issue https://github.com/zephyrproject-rtos/zephyr/issues/20991

FRASTM commented:
When the LPTIM is needed, the  
CONFIG_STM32_LPTIM_TIMER=y
must be set

FRASTM committed:
driver: timer: stm32_lptim: tickless mode without a compilation flag

In the timeout function, remove the compilation flag
and use the macro instead.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
driver: timer: stm32_lptim set_timeout limit values

set the min and max values of the given ticks from 0
to LPTIM_TIMEBASE which is the full register value
In case the timeout is FOREVER, then lptimer is stopped

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
driver: timer: stm32_lptim control the timer duration

This change makes the lptimer running with lower tick periods
and small tick values

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
soc: arm: stm32 TICKS value with LPTIM config

This is defining the SYS_CLOCK_TICKS_PER_SEC default value
depending on the LPTIM CLOCK frequency in case of LPTIMER,
to get a TICK value as a divider of the LPTIM clock source.
It gives a better result in formulas when converting
ticks to count unit.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: stm32: timer define a new DT binding for LPTIM

Adds a new LPTIM binding for stm32 soc,
based on the timer binding. This will makes a specific filter
on dt_compat_enabled("stm32,lptim")

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Takes the https://github.com/zephyrproject-rtos/zephyr/pull/25126
Configures some tests/kernel   testcases to include the LPTIM feature
Add a DT for the stm32-lptim yaml

erwango commented on dts/bindings/timer/st,stm32-lptim.yaml:
```diff
@@ -0,0 +1,5 @@
+description: STM32 lptim
```
Please put the usual header:
```
# Copyright (c) 2020, STMicroelectronics
# SPDX-License-Identifier: Apache-2.0
```

FRASTM replied:
added also in the dts/bindings/timer/st,stm32-timers.yaml

erwango commented on soc/arm/st_stm32/common/Kconfig.defconfig.series:
```diff
@@ -22,6 +22,10 @@ config SYS_CLOCK_TICKS_PER_SEC
 	default 4096
 	depends on STM32_LPTIM_TIMER && STM32_LPTIM_CLOCK_LSE
 
+config TICKLESS_KERNEL
```
This should not be need as STM32_LPTIM_TIMER is declared TICKLESS_CAPABLE, and TICKLESS_KERNEL is declared in the following way:
```
config TICKLESS_KERNEL
	bool "Tickless kernel"
	default y if TICKLESS_CAPABLE
```
So the only protection needed is a check in driver to return an error if not support which is also proposed in this PR.
I think we can remove this commit safely.

FRASTM replied:
removed

FRASTM commented:
Testing on nucleo_wb55rg with CONFIG_STM32_LPTIM_CLOCK_LSE and CONFIG_SYS_CLOCK_TICKS_PER_SEC=4096
--> tests/kernel/timer and tests/kernel/early_sleep PASSED
--> tests/kernel/sched/schedule_api/ still fails if CONFIG_TIMESLICING=y

erwango commented:
@jdascenzio would you mind review this PR ?
If you agree your commits would be merged as part of this PR.

FRASTM committed:
boards: arm: stm32l4r5 unset the LPTIM clock source if enabled

For nucleo_l4r5zi board, the LPTIM clock source is defined
as LSE as default.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

jdascenzio commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -274,15 +310,30 @@ u32_t z_timer_cycle_get_32(void)
 	lp_time += accumulated_lptim_cnt;
 
+	/* convert lptim count in a nb of hw cycles with precision */
+#if defined(CONFIG_STM32_LPTIM_CLOCK_LSI)
+	u64_t ret = lp_time * ((sys_clock_hw_cycles_per_sec() / 1000)
+		/ (LPTIM_CLOCK / 1000));
+#else
+	u64_t ret = lp_time * (sys_clock_hw_cycles_per_sec() / LPTIM_CLOCK);
+#endif /* CONFIG_STM32_LPTIM_CLOCK_LSI */
```
Why it's different for LSI?
Here, I think you want to do:
```u64_t ret = (lp_time * sys_clock_hw_cycles_per_sec()) / LPTIM_CLOCK;```

FRASTM replied:
simplification to have the same return value in case of LSI or LSE

jdascenzio commented on soc/arm/st_stm32/common/Kconfig.defconfig.series:
```diff
@@ -9,8 +9,18 @@ if SOC_FAMILY_STM32
+	depends on STM32_LPTIM_TIMER && STM32_LPTIM_CLOCK_LSI
+
+# set the tick per sec as a divider of the LPTIM clock source
+config SYS_CLOCK_TICKS_PER_SEC
+	default 4096
+	depends on STM32_LPTIM_TIMER && STM32_LPTIM_CLOCK_LSE
```
Like @aurel32 said (https://github.com/zephyrproject-rtos/zephyr/pull/25126#issuecomment-627639520), it could be interesting to have 32768 here instead of 4096

FRASTM replied:
Yes, but this value of 32768 for TICKS_PER_SEC is too high and cannot pass the kernel/tests timers_api

FRASTM replied:
testing kernel/tests on nucleo_l4r5zi with
CONFIG_STM32_LPTIM_TIMER=y1
CONFIG_SYS_CLOCK_TICKS_PER_SEC=32768

jdascenzio commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -226,6 +235,16 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
+	 * If this event got outdated fire interrupt right now,
+	 * else schedule it normally.
+	 */
+	if (next_arr <= ((LL_LPTIM_GetCounter(LPTIM1) + 1) & LPTIM_TIMEBASE)) {
+		NVIC_SetPendingIRQ(LPTIM1_IRQn);
+		lptim_fired = 1;
+	}
```
LPTIM1 current value must be read like this:
```
	while (lp_time != LL_LPTIM_GetCounter(LPTIM1)) {
		lp_time = LL_LPTIM_GetCounter(LPTIM1);
	}
```

FRASTM replied:
changed LL_LPTIM_GetCounter  with lp_time value as this variable is already set.

jdascenzio commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -50,6 +51,13 @@ static void lptim_irq_handler(struct device *unused)
 		/* do not change ARR yet, z_clock_announce will do */
 		LL_LPTIM_ClearFLAG_ARRM(LPTIM1);
 
+		if (lptim_fired) {
+			lptim_fired = 0;
+#if defined(LPTIM_CR_COUNTRST)
+			LL_LPTIM_ResetCounter(LPTIM1);
+#endif
+		}
+
```
When lptim_fired, I think that LL_LPTIM_IsActiveFlag_ARRM could be not set...

FRASTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -251,12 +270,29 @@ u32_t z_clock_elapsed(void)
+#endif /* CONFIG_STM32_LPTIM_CLOCK_LSI */
 
 	/* convert to ticks */
 	return z_ms_to_ticks(ret);
```
correct

FRASTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -251,12 +270,29 @@ u32_t z_clock_elapsed(void)
 	 * since the previous 'announce'
 	 */
-	u32_t ret = ((lp_time + 1) * 1000) / LPTIM_CLOCK;
+#if defined(CONFIG_STM32_LPTIM_CLOCK_LSI)
+	u32_t ret = (lp_time * (CONFIG_SYS_CLOCK_TICKS_PER_SEC / 1000))
+		/ (LPTIM_CLOCK / 1000);
+
+	return (u32_t)(ret);
+#else
+	u64_t ret = (lp_time * CONFIG_SYS_CLOCK_TICKS_PER_SEC) / LPTIM_CLOCK;
+
+	return (u32_t)(ret);
+#endif /* CONFIG_STM32_LPTIM_CLOCK_LSI */
```
just a question of type, which could be simplified with u64_t in all cases

FRASTM commented:
running testcases  kernel/tests on nucleo_l4r5zi with
CONFIG_STM32_LPTIM_TIMER=y1
CONFIG_SYS_CLOCK_TICKS_PER_SEC=32768

FRASTM committed:
drivers: timer: stm32_lptim fix TICKLESS=n processing

based on PR#25412
Some kernel tests use `CONFIG_TICKLESS_KERNEL=n` with
`CONFIG_SYS_CLOCK_TICKS_PER_SEC=1` to detect when a test runs longer
than 1 second.  These tests break if a tick is announced every time a
timeout occurs.  Only announce if the measured duration since the last
tick is at least the duration of a tick.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -193,17 +193,8 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
-		ticks = 0xFFFF;
-	}
+	ticks = (ticks == K_TICKS_FOREVER) ? LPTIM_TIMEBASE : ticks;
+	ticks = MAX(MIN(ticks - 1, (s32_t)LPTIM_TIMEBASE), 1);
```
correct, 
if TICKS ==  K_TICKS_FOREVER  lptim has been stopped just before.
Since the LL_LPTIM_Disable does not exist, the LPTIM clock is stopped.

FRASTM commented:
adding the https://github.com/zephyrproject-rtos/zephyr/pull/25412  in case of tests with CONFIG_TICKLESS_KERNEL=n
Test with nucleo_l4r5zi with LPTIM sourced by LSE and TICKS_PER_SEC=4096

carlescufi commented:
@erwango could you take a look at this one?

FRASTM commented:
Note that test/kernel/interrupt 
still few tests fail on l4r5zi with LSE clock source and TICKS_PER_SEC=4096
tests/kernel/common/kernel.common   
tests/kernel/sleep/   
tests/kernel/fifo/fifo_timeout/kernel.fifo.timeout

Note that tests/kernel/interrupt  could fail due to the irq pemption on lptim irq itself

carlescufi commented:
@erwango @jdascenzio can we move this forward so we can get the fix in in time for 2.3.0? Thanks!

FRASTM committed:
drivers: timer: lptim is [EXPERIMENTAL] for stm32 soc series only

Activation of the LPTIMER is valid for SLEEP MODE only
The choice of the lptim clock source is STM32_LPTIM_CLOCK
set the LSE in first position to have as default value

Signed-off-by: Francois Ramu <francois.ramu@st.com>

carlescufi commented:
@erwango can you take another look at the PR after @FRASTM pushed an update?

andyross reviewed (approved):
Late to this one, sorry.  I can't speak to the hardware obviously, but the timer logic all looks correct.

erwango commented on soc/arm/st_stm32/common/Kconfig.defconfig.series:
```diff
@@ -9,8 +9,16 @@ if SOC_FAMILY_STM32
+	default y
+	depends on !STM32_LPTIM_TIMER
+
+# set the tick per sec as a divider of the LPTIM clock source
+config SYS_CLOCK_TICKS_PER_SEC
+	default 4096 if STM32_LPTIM_TIMER && STM32_LPTIM_CLOCK_LSE
+
+# set the tick per sec as a divider of the LPTIM clock source
+config SYS_CLOCK_TICKS_PER_SEC
+	default 4000 if  STM32_LPTIM_TIMER && STM32_LPTIM_CLOCK_LSI
```
Please make this as follows:
```
# set the tick per sec as a divider of the LPTIM clock source
config SYS_CLOCK_TICKS_PER_SEC
	default 4096 if STM32_LPTIM_TIMER && STM32_LPTIM_CLOCK_LSE
	default 4000 if STM32_LPTIM_TIMER && STM32_LPTIM_CLOCK_LSI
```

erwango commented on drivers/timer/Kconfig.stm32_lptim:
```diff
@@ -17,26 +17,26 @@ if STM32_LPTIM_TIMER
 config STM32_LPTIM_CLOCK
 	int "LPTIM clock value"
-	default 32000 if STM32_LPTIM_CLOCK_LSI
 	default 32768 if STM32_LPTIM_CLOCK_LSE
+	default 32000 if STM32_LPTIM_CLOCK_LSI
```
Not useful but ok

erwango commented on drivers/timer/Kconfig.stm32_lptim:
```diff
@@ -17,26 +17,26 @@ if STM32_LPTIM_TIMER
 config STM32_LPTIM_TIMEBASE
 	hex "LPTIM AutoReload value"
-	default 0xF9FF if STM32_LPTIM_CLOCK_LSI
 	default 0xFFFF if STM32_LPTIM_CLOCK_LSE
+	default 0xF9FF if STM32_LPTIM_CLOCK_LSI
```
Not useful but ok

FRASTM replied:
setting LSE in first position to have this as default ?

erwango replied:
Position only matters for choice (like STM32_LPTIM_CLOCK) actually

erwango commented on boards/arm/nucleo_l4r5zi/Kconfig.defconfig:
```diff
@@ -9,7 +9,7 @@ config BOARD
 	default "nucleo_l4r5zi"
 
 choice STM32_LPTIM_CLOCK
-	default STM32_LPTIM_CLOCK_LSI
+	default STM32_LPTIM_CLOCK_LSE
 	depends on STM32_LPTIM_TIMER
 endchoice
```
I think we could completely remove this choice section.

FRASTM replied:
yes, removed

FRASTM replied:
yes, removed

erwango commented:
@jdascenzio, are you fine with the last changes ?

FRASTM commented:
change after @erwango 's review

erwango commented on dts/arm/st/l4/stm32l4.dtsi:
```diff
@@ -340,7 +340,7 @@
 
 		lptim1: timers@40007c00 {
-			compatible = "st,stm32-timers";
+			compatible = "st,stm32-lptim";
```
Can you fix this on stm32wb as well ?

FRASTM replied:
yes

erwango reviewed (changes requested):
Few more tweaks and I'm good

FRASTM commented:
rebase

aurel32 commented on drivers/timer/Kconfig.stm32_lptim:
```diff
@@ -4,8 +4,8 @@
-	bool "STM32 Low Power Timer"
-	depends on (SOC_SERIES_STM32L4X || SOC_SERIES_STM32WBX)
+	bool "STM32 Low Power Timer [EXPERIMENTAL]"
+	depends on SOC_SERIES_STM32L4X && DEVICE_POWER_MANAGEMENT
```
Minor nitpick: I don't fully see the relation of why the LPTIM support needs to depend on DEVICE_POWER_MANAGEMENT.

FRASTM replied:
My goal is to link lptim to low power mode. 
This let kernel timer unchanged for most of the usecase.
The lptim is able to preserve the kernel tick counter during low power sleep mode : only in this case is the lptim configured to keep the kernel timeout. 
Refering to  power: stm32l4: support of the Low Power Mode #19026

aurel32 replied:
Understood. In that case shouldn't it be `SYS_POWER_MANAGEMENT` instead of `DEVICE_POWER_MANAGEMENT`?

aurel32 reviewed (approved):
Beside the minor nitpick, all looks good. I tested it on a Nucleo L432Kc board and I confirm the issue is fixed by this PR.

aurel32 commented:
Another comment: when setting `SYS_CLOCK_TICKS_PER_SEC=32768` (instead of 4096) with a LSI OSC, `test_timer_remaining` fails.

This seems to be a math issue in `k_busy_wait` and setting both `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` and `SYS_CLOCK_TICKS_PER_SEC` to 32768 (like done on the nRF52 series) just makes the issue to disappear.

carlescufi commented:
@erwango please re-review

FRASTM commented:
include lptim in  the stm32wbx series
